internal/appsec/waf/waf.go

```diff
@@ -516,6 +574,74 @@ func (e *encoder) encodeUint64(n uint64, wo *wafObject) error {
 	return e.encodeString(strconv.FormatUint(n, 10), wo)
 }
 
+func decodeObject(wo *wafObject) (v interface{}, err error) {
+	if wo == nil {
+		return nil, errBadWafObjectPtr
+	}
+	switch wo._type {
+	case wafUintType:
+		return uint64(*wo.uint64ValuePtr()), nil
+	case wafIntType:
+		return int64(*wo.int64ValuePtr()), nil
+	case wafStringType:
+		return C.GoString(*wo.stringValuePtr()), nil
+	case wafArrayType:
+		return decodeArray(wo)
+	case wafMapType: // could be a map or a struct, no way to differentiate
+		return decodeMap(wo)
+	default:
+		return nil, errUnsupportedValue
+	}
+}
+
+func decodeArray(wo *wafObject) ([]interface{}, error) {
+	if wo == nil {
+		return nil, errBadWafObjectPtr
+	}
+	var err error
+	len := wo.length()
+	arr := make([]interface{}, len)
+	for i := C.uint64_t(0); i < len && err == nil; i++ {
+		arr[i], err = decodeObject(wo.index(i))
+	}
+	return arr, err
+}
+
+func decodeMap(wo *wafObject) (map[string]interface{}, error) {
+	if wo == nil {
+		return nil, errBadWafObjectPtr
+	}
+	length := wo.length()
+	decodedMap := make(map[string]interface{}, length)
+	for i := C.uint64_t(0); i < length; i++ {
+		obj := wo.index(i)
+		if key, err := decodeMapKey(obj); err == nil {
+			if val, err := decodeObject(obj); err == nil {
+				decodedMap[key] = val
+			} else {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
```

Comment:
IMHO, more readable with:
```suggestion
		key, err := decodeMapKey(obj)
		if err != nil {
			return nil, err
		}
		val, err := decodeObject(obj)
		if err != nil {
			return nil, err
		}
		decodedMap[key] = val
```

Reply:
Nice, thanks